Documentation: expansion of Monkeypatch to include mocked classes and dictionaries (#5315)
pytest-dev/pytest: doc/en/monkeypatch.rst

on global settings or which invokes code which cannot be easily
tested such as network access.  The ``monkeypatch`` fixture
helps you to safely set/delete an attribute, dictionary item or
environment variable, or to modify ``sys.path`` for importing.

The ``monkeypatch`` fixture provides these helper methods for safely patching and mocking
functionality in tests:

.. code-block:: python

    monkeypatch.setattr(obj, name, value, raising=True)
    monkeypatch.delattr(obj, name, raising=True)
    monkeypatch.setitem(mapping, name, value)
    monkeypatch.delitem(obj, name, raising=True)
    monkeypatch.setenv(name, value, prepend=False)
    monkeypatch.delenv(name, raising=True)
    monkeypatch.syspath_prepend(path)
    monkeypatch.chdir(path)

All modifications will be undone after the requesting
test function or fixture has finished. The ``raising``
parameter determines if a ``KeyError`` or ``AttributeError``

> will be raised if the set/deletion operation has no target.

Reviewer: ```suggestion
will be raised if the target of the set/deletion operation does not exist.
```

Author: Applied in https://github.com/pytest-dev/pytest/commit/7fa0e0405d4c051d176ba40ae4205dc3393e27a7

Consider the following scenarios:

1. Modifying the behavior of a function or the property of a class for a test e.g.
there is an API call or database connection you will not make for a test but you know
what the expected output should be. Use :py:meth:`monkeypatch.setattr` to patch the
function or property with your desired testing behavior. This can include your own functions.
Use :py:meth:`monkeypatch.delattr` to remove the function or property for the test.

2. Modifying the values of dictionaries e.g. you have a global configuration that
you want to modify for certain test cases. Use :py:meth:`monkeypatch.setitem` to patch the
dictionary for the test. :py:meth:`monkeypatch.delitem` can be used to remove items.

3. Modifying environment variables for a test e.g. to test program behavior if an
environment variable is missing, or to set multiple values to a known variable.
:py:meth:`monkeypatch.setenv` and :py:meth:`monkeypatch.delenv` can be used for
these patches.

> 4. Use :py:meth:`monkeypatch.syspath_prepend` to add to the system ``$PATH`` and

Reviewer: ```suggestion
4. Use :py:meth:`monkeypatch.syspath_prepend` to modify ``sys.path`` safely, and
```

Author: Applied in https://github.com/pytest-dev/pytest/pull/5315/commits/60e4fd448e0ed05bd71436e506fafd4643b0a105

:py:meth:`monkeypatch.chdir` to change the context of the current working directory
during a test.

See the `monkeypatch blog post`_ for some introduction material
and a discussion of its motivation.

.. _`monkeypatch blog post`: http://tetamap.wordpress.com/2009/03/03/monkeypatching-in-unit-tests-done-right/

Simple example: monkeypatching functions
----------------------------------------

Consider a scenario where you are working with user directories. In the context of

> testing, you do not want your test to depend on the running user. ``Monkeypatch``

Reviewer 2: I think we should consistently match the case of the fixture name, so

```rst
    ``monkeypatch``
```

instead of

```rst
    ``Monkeypatch``
```

can be used to patch functions dependent on the user to always return a
specific value.

In this example, :py:meth:`monkeypatch.setattr` is used to patch ``os.path.expanduser``
so that the known testing string ``"/abc"`` is always used when the test is run.
This removes any dependency on the ``$USER`` environment variable for testing purposes.
:py:meth:`monkeypatch.setattr` must be called before the function which will use
the patched function is called.
After the test function finishes the ``os.path.expanduser`` modification will be undone.

.. code-block:: python

    # Content of test_module.py with source code and the test.
    # os.path is imported for reference in monkeypatch.setattr().
    import os.path

    def getssh():
        """Simple function to return expanded homedir ssh path from $USER"""

>         user = os.getenv("$USER")

Reviewer 2: this looks up the environment variable named '$USER` -- I think you mean `os.getenv('USER')` (which could return `None`)

The `os.path.expanduser` call for "the current user" would just be:

```python
    return os.path.join(os.path.expanduser("~"), ".ssh")
```

no need for `getenv` or string formatting

>         return os.path.join(os.path.expanduser(f"~{user}"), ".ssh")

Reviewer: To simplify this example, you can use:

```python
def getssh():
    """Simple function to return expanded homedir ssh path from $USER"""
    return os.path.expanduser("~/.ssh")
```

Author: Applied in https://github.com/pytest-dev/pytest/commit/7fa0e0405d4c051d176ba40ae4205dc3393e27a7 - this is a really good simplification. I was debating using `Path.home()` but the original example used `os` so I wanted to stick with it.

        # Any arguments may be passed and it will always return our
        # mocked object, which only has the .json() method.
        def mock_get(*args, **kwargs):
            return MockResponse()

        # Apply the monkeypatch for requests.get to mock_get.
        monkeypatch.setattr(requests, "get", mock_get)

        # Our simple but illustrative test.
        # app.get_json, which contains requests.get, uses the monkeypatch.
        # Therefore, our result is our mock.
        result = app.get_json("https://fakeurl")
        assert result["mock_key"] == "mock_response"

``Monkeypatch`` applies the mock for ``requests.get`` with our ``mock_get`` function.
The ``mock_get`` function returns an instance of the ``MockResponse`` class, which
has a ``json()`` method defined to return a known testing dictionary and does not
require any outside API connection.

You can build the ``MockResponse`` class with the appropriate degree of complexity for
the scenario you are testing. For instance, it could include an ``ok`` property that
always returns ``True``, or return different values from the ``json()`` mocked method
based on input strings.

> This mock can be shared across tests using the ``fixture`` syntax:

Reviewer: ```suggestion
This mock can be shared across tests using a ``fixture``:
```

A fixture is not really a new syntax. 😁

Author: Applied in https://github.com/pytest-dev/pytest/commit/7fa0e0405d4c051d176ba40ae4205dc3393e27a7

.. code-block:: python

    # Contents of test_app.py, a simple test for our API retrieval.
    import pytest
    import requests

    # Our app.py that includes the get_json() function.
    import app

    # A Custom class to be the mock.
    class MockResponse:
        @staticmethod
        def json():
            return {"mock_key": "mock_response"}

    # Monkeypatched requests.get moved to a fixture.
    @pytest.fixture
    def mock_response(monkeypatch):
        """Requests.get() mocked to return {'mock_key':'mock_response'}."""

        def mock_get(*args, **kwargs):
            return MockResponse()

        monkeypatch.setattr(requests, "get", mock_get)

    # Notice our test uses the custom fixture instead of monkeypatch directly.
    def test_get_json(mock_response):
        result = app.get_json("https://fakeurl")
        assert result["mock_key"] == "mock_response"

Furthermore, if the mock was designed to be applied to all tests, the ``fixture`` could

> be moved to ``conftest.py`` with ``autouse=True``.

Reviewer: ```suggestion
be moved to a ``conftest.py`` file and use the ``autouse=True`` option.
```

Author: Applied in https://github.com/pytest-dev/pytest/commit/7fa0e0405d4c051d176ba40ae4205dc3393e27a7

.. code-block:: python

>     # Contents of our original code file e.g. code.py.

Reviewer: I would leave all those "contents -> Contents" changes lone, just for consistency with the rest of the docs.

Author: Great catch, also ensured `monkeypatch` is lowercase everywhere except when it is the first word in the headers.

Reviewer: I still see comments starting with an upper case letter... would you mind changing all comments to start with lowercase? This is the pattern we use through the docs and code. 👍

Reviewer 3: > I still see comments starting with an upper case letter... would you mind changing all comments to start with lowercase? This is the pattern we use through the docs and code. 

That's new to me also - I see that there are comments without punctuation and starting lowercased, but when using punctuation / full sentences they should be capitalized (like above).

Reviewer: full sentences yes, but in one-liners like this we don't use punctuation and start with a lowercase letter (at least that's what I see browsing the docs quickly now)

Reviewer 3: Well, you will also see no typing hints currently - but maybe soon.. ;)
(what I want to say is that style can be improved, and I prefer it like above also)

Reviewer: Well I won't get into that discussion right now then, the current style is fine.

@EvanKepner could you just squash and rebase your commits then? Thanks!

Author: No problem at all, squash/rebase coming soon. Also @nicoddemus sorry I misread your original comment and thought you wanted all comment lines capitalized but without punctuation unless part of a larger sentence :sweat_smile: 

I'm a fan of generally strict type hints and full doc-strings for each function. It's a difficult challenge with documentation because you don't want to add additional syntax that could confuse a reader. These tutorial examples are simplified to show how the patching works so I steered towards the least verbose form. 

Might be worth an update in the guidance to https://github.com/pytest-dev/pytest/blob/master/CONTRIBUTING.rst#write-documentation if you come to a consensus on standardized documentation style beyond the pre-commit checks for `black` in the docs. If you're seeking suggestions I like one sentence per line in RST with a hard-wrap (or just a hard-wrap at something sensible) to make suggested changes and PR reviews easier.

.. code-block:: python

>     # Contents of our test file e.g. test_code.py.

Reviewer: Also, if I'm being pedantic, we usually don't use `.` at the end of single line comments

Author: I changed this throughout the file with regard to the single line comments.

    import app

    def test_connection(monkeypatch):

        # Patch the values of DEFAULT_CONFIG to specific
        # testing values only for this test.
        monkeypatch.setitem(app.DEFAULT_CONFIG, "user", "test_user")
        monkeypatch.setitem(app.DEFAULT_CONFIG, "database", "test_db")

        # The expected result based on the mocks.
        expected = "User Id=test_user; Location=test_db;"

        # The test uses the monkeypatched dictionary settings.
        result = app.create_connection_string()
        assert result == expected

You can use the :py:meth:`monkeypatch.delitem` to remove values.

.. code-block:: python

    # Contents of test_app.py.
    import pytest

    # Our app.py with the connection string function.
    import app

    def test_missing_user(monkeypatch):

        # Patch the DEFAULT_CONFIG t be missing the 'user' key.
        monkeypatch.delitem(app.DEFAULT_CONFIG, "user", raising=False)

        # Key error expected because a config is not passed, and the
        # default is now missing the 'user' entry.
        with pytest.raises(KeyError):
            _ = app.create_connection_string()

> The modularity of the ``fixture`` syntax gives you the flexibility to define

Reviewer: ```suggestion
The modularity of fixtures gives you the flexibility to define
```

Author: Applied in https://github.com/pytest-dev/pytest/commit/7fa0e0405d4c051d176ba40ae4205dc3393e27a7 - also realized the second test does not require the `mock_test_database` fixture so I removed it from the signature - https://github.com/pytest-dev/pytest/commit/7fa0e0405d4c051d176ba40ae4205dc3393e27a7#diff-425daef754438812ec1a0e4d3aeaeff8L434
